#  4.7: Deriving New Variables

## Table of content:

### Task 1 - If you haven’t done so already, complete the instructions in the Exercise for creating the “price_label” and “busiest_day” columns.>

#### 1) If-Statements with User-Defined Functions

#### 2) If-Statements with the loc() Function

### Task 2 - Suppose your clients have changed their minds about the labels you created in your “busiest_day” column. Now, they want “Busiest day” to become “Busiest days” (plural). This label should correspond with the two busiest days of the week as opposed to the single busiest day. At the same time, they’d also like to know the two slowest days. Create a new column for this using a suitable method.

### Task 3 - Check the values of this new column for accuracy. Note any observations in markdown format.

### Task 4 - When too many users make Instacart orders at the same time, the app freezes. The senior technical officer at Instacart wants you to identify the busiest hours of the day. Rather than by hour, they want periods of time labeled “Most orders,” “Average orders,” and “Fewest orders.” Create a new column containing these labels called “busiest_period_of_day.”

### Task 5 - Print the frequency for this new column.

### Task 6 - Ensure your notebook is clean and structured and that your code is well commented.

### Task 7 - Export your dataframe as a pickle file (since you added new columns) and store it correctly in your “Prepared Data” folder.



#### Import Libraries

In [1]:
# Import libraries

import pandas as pd
import numpy as np
import os

#### Import Data Set

In [2]:
path = r'C:\Users\facun\Desktop\Data Analysis\CF\PYTHON\Instacart Basket Analysis'

In [43]:
# Import df_ords_merged dataframe

ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_merged.pkl'))

## Task 1:

#### If you haven’t done so already, complete the instructions in the Exercise for creating the “price_label” and “busiest_day” columns.

### 1) If-Statements with User-Defined Functions

In [33]:
# Define function

def add_num(x,y):         # definition - descriptive name - two arguments
    sum = x + y           # }Body: new variable - two arguments
    return sum            # printing: sum is returned

In [35]:
# Run Function

add_num(9,11)

20

#### Sort your products into price ranges groups by creating `flags`

Apply the following criteria: 

- If the item’s price is lower than or equal to $5, it will be labeled a “low-range product.”

- If the item’s price is above USD 5 but lower than or equal to USD 15, it will be labeled a “mid-range product.”

- If the item’s price is above USD 15, it will be labeled a “high-range product.”

#### Create a subset to avoid any potential issues that can lead to memory issues or processing power because of the larga dataframe

In [44]:
df = ords_prods_merge[:1000000] # Everything from beginning up to the number

In [1]:
df.shape

NameError: name 'df' is not defined

#### Define a function to sort data by categories

In [31]:
df.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_last_order  product_id  add_to_cart_order  reordered  \
0                    NaN         196                  1          0   
1                   15.0         196                  1          1   
2                   21.0         196                  1          1   
3                   29.0         196                  1          1   
4                   28.0         196                  1          1   

  product_name  aisle_id  department_id  prices _merge        price_range  
0         Soda        77              7     9.0   both  Mid-range product  
1         Soda        77              7     9.0   both  Mid-range product  
2         Soda        77              7     9.0   both  Mid-range product  
3         Soda        77              7     9.0   both  Mid-range product  
4         Soda        77              7     9.0   both  Mid-range product

In [26]:
def price_label(row): # creating new function - name of the function - argument (row)

  if row['prices'] <= 5:                              # If the item’s price is lower than or equal to $5
    return 'Low-range product'                        # it will be labeled a “low-range product
  elif (row['prices'] > 5) and (row['prices'] <= 15): # If the item’s price is above $5 but lower than or equal to $15
    return 'Mid-range product'                        # it will be labeled a “mid-range product.”
  elif row['prices'] > 15:                            # If the item’s price is above $15
    return 'High range'                               # it will be labeled a “high-range product.”
  else: return 'Not enough data'                      # If any of the above can't be found then return "Not enough data"

#### Apply functions


**`df['price_range']`** : cretates new column called **price_range** designated as location for price_label

**`df.apply(price_label, axis=1)`** : tells Python to **apply** the **price_label** function on **axis=1**

**axis=1** stands for "**row**" so this code essentially tells Python to apply the function to all **`rows`** within the dataframe.

**axis=0** Would tell Python to apply the function to all **`columns`** within the dataframe

In [27]:


df['price_range'] = df.apply(price_label, axis=1)

C:\Users\facun\AppData\Local\Temp\ipykernel_23600\2848743372.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['price_range'] = df.apply(price_label, axis=1)


#### Check values in you column:

**`value_counts()`**

In [28]:
df['price_range'].value_counts(dropna = False)

Mid-range product    756450
Low-range product    243550
Name: price_range, dtype: int64

Outcome:

- There is NONE products above $15
- There are 756,450 products between \$ 5 and equal to \\$15 
- There are 243,550 products with prices lower than \$5

#### Check the most expensive product within the subset:

In [30]:
df['prices'].max()

14.8

Outcome:

The result comfirms the finding as there is no product labelled $15 as the maximum value within "prices" column is \\$14.8

In [32]:
df.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_last_order  product_id  add_to_cart_order  reordered  \
0                    NaN         196                  1          0   
1                   15.0         196                  1          1   
2                   21.0         196                  1          1   
3                   29.0         196                  1          1   
4                   28.0         196                  1          1   

  product_name  aisle_id  department_id  prices _merge        price_range  
0         Soda        77              7     9.0   both  Mid-range product  
1         Soda        77              7     9.0   both  Mid-range product  
2         Soda        77              7     9.0   both  Mid-range product  
3         Soda        77              7     9.0   both  Mid-range product  
4         Soda        77              7     9.0   both  Mid-range product

### 2) If-Statements with the loc() Function

Using **loc()**, you can apply the conditional logic of an if-statement to a function without explicitly creating an **if-els**e construct.

#### Create conditions as the previous method

In [36]:
# 1)

df.loc[df['prices'] > 15, 'price_range_loc'] = 'High-range product'

C:\Users\facun\AppData\Local\Temp\ipykernel_23600\1169838859.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.loc[df['prices'] > 15, 'price_range_loc'] = 'High-range product'


In [38]:
# 2)

df.loc[(df['prices'] <= 15) & (df['prices'] > 5), 'price_range_loc'] = 'Mid-range product' 

In [50]:
# 3) 

 
df.loc[df['prices'] <= 5, 'price_range_loc'] = 'Low-range product'

In [40]:
df['price_range_loc'].value_counts(dropna = False) # value.counts(): counts number o products within each label 

Mid-range product    756450
Low-range product    243550
Name: price_range_loc, dtype: int64

#### 1)

- **if** = `df.loc[df['prices'] > 15,`

- **then** = `'price_range_loc'] = 'High-range product'`


#### 2)

- **if** = `df.loc[(df['prices'] <= 15) `**&**` (df['prices'] > 5),` These two conditions are combined by the **&** sign in the middle of the “if” half.
- **then** = `'price_range_loc'] = 'Mid-range product'`


#### 3)

- **if** = `df.loc[df['prices'] <= 5,`

- **then** = `'price_range_loc'] = 'Low-range product'`

### Why loc() over User-defined-function?


1. Using **loc()** won’t result in a warning message

2. **loc()** method runs much faster; the **loc()** function applies the conditional filters before searching through the dataframe, while your **user-defined function** searches through the entire dataframe and then determines where to set the filters (remember **axis = 1**?).

### Repeat process with the entire data frame

In [47]:
ords_prods_merge.loc[ords_prods_merge['prices'] > 15, 'price_range_loc'] = 'High-range product'

In [48]:
ords_prods_merge.loc[(ords_prods_merge['prices'] <= 15) & (df['prices'] > 5), 'price_range_loc'] = 'Mid-range product' 

In [51]:
ords_prods_merge.loc[ords_prods_merge['prices'] <= 5, 'price_range_loc'] = 'Low-range product'

In [53]:
ords_prods_merge['price_range_loc'].value_counts(dropna = False)

NaN                   21105547
Low-range product     10126366
Mid-range product       756450
High-range product      417678
Name: price_range_loc, dtype: int64

### 3) If-Statements with For-Loops

**`For-loops`**, are **loops** for running the same block of **code multiple times**. Perform the same function on multiple elements, for instance, by running through an entire dataframe and performing a function on each row within that dataframe.

Example:

Print the statement **“My age is X”** for ages 30 to 45 (and substituting the age in question for “X” for each one). Without a **for-loop**, manually print statements. 

In [54]:
print("My age is 30")

My age is 30


In [55]:
print("My age is 31")

My age is 31


In [56]:
print("My age is 32")

My age is 32


In [57]:
print("My age is 33")

My age is 33


#### With for-loop

Yu can print everything with a single command

In [59]:
for x in range(30,45):
    print("My age is %d" % (x)) 

My age is 30
My age is 31
My age is 32
My age is 33
My age is 34
My age is 35
My age is 36
My age is 37
My age is 38
My age is 39
My age is 40
My age is 41
My age is 42
My age is 43
My age is 44


#### Function breakdown

The **loop** cycles for every value **(x)** within the range of 30 and 45. Each time it cycles, it executes the code inside the loop, which, in this case, prints `“My age is”` followed by the current value of x

**"My age is %d"** is a string that includes a placeholder **%d**, which will be replaced by the value of the variable **x**.
**`% (x)`** is used to format the string, replacing **%d** with the current value of **x**.

### Create a new column in the ords_prods_merge dataframe that summarizes thow busy each day of the week is

In [62]:
# 1) : First we need to know on which day most orders take place by printing the frequency of the “orders_day_of_the_week” column

ords_prods_merge['orders_day_of_week'].value_counts(dropna = False)

0    6204404
1    5660456
6    4496635
2    4213986
5    4205906
3    3840701
4    3783953
Name: orders_day_of_week, dtype: int64

Brief: 

0 = Saturday
1 = Sunday
2 = Monday
3 = Tuesday
4 = Wednesday
5 - Thursday
6 - Friday

Result breakdown:

**0** has the highest frequency, meaning **Saturday** is the busiest day and **4**, **Wednesday**, is the slowest day of orders

We need the information to create a new column, **busiest day** that will contain: "Busiest day", "Least busy", and "Regurlarly busy". 

In [65]:
result = []                                         # Create an empty list " result"

for value in ords_prods_merge["orders_day_of_week"]:# For each value within the column "orders of week" we want something
  if value == 0:                                    # if the value of the row is equal to 0
    result.append("Busiest day")                    # then "Busiest day" is appended to the blank "result" list
  elif value == 4:                                  # if the value is equal to 4
    result.append("Least busy")                     # then " Least busy" is appended to the "result" list
  else:                                             # if is neither of these conditions are met
    result.append("Regularly busy")                 # then 'Regularly busy'

# "value" acts as placeholder, it could be called anything, could be also an X. 
# Here represents every possible age that could exist between 30 and 45. 
# Here represents "orders_day_of_week"

In [67]:
# After running loop, print gthe result list

result

['Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Busiest day',
 'Regularly busy',
 'Reg

The long list procuded does not do anything by itself. We need to combine it with **`ords_prods_merge`** datafame to can be used effectively. 

To combine it, we need to create a new column within **`ords_prods_merge`** and **`result`**

In [68]:
ords_prods_merge['businest_day'] = result

In [69]:
# Print the frequency of the new column and cross-check it with the frequency of "orders_day_of_week"

ords_prods_merge['businest_day'].value_counts(dropna = False)

Regularly busy    22417684
Busiest day        6204404
Least busy         3783953
Name: businest_day, dtype: int64

With that, you’ve successfully created a summary column for the busiest day of the week!

## Task 2:

Suppose your clients have changed their minds about the labels you created in your “busiest_day” column. Now, they want “Busiest day” to become “Busiest days” (plural). This label should correspond with the ***`two busiest days`*** of the week as opposed to the single busiest day. At the same time, they’d also like to know the ***`two slowest days`***. Create a new column for this using a suitable method.

In [70]:
# 1) We check which are the 2 busiest days and the 2 slowest days

ords_prods_merge['orders_day_of_week'].value_counts(dropna = False)

0    6204404
1    5660456
6    4496635
2    4213986
5    4205906
3    3840701
4    3783953
Name: orders_day_of_week, dtype: int64

Answer: 0 and 1 (Saturday and Sunday) are the Busiest days and the slowest days are 3 and 4 (Tuesday and Wednesday)

In [113]:
# 2) 

result = []                                         # Create an empty list " result"

for value in ords_prods_merge["orders_day_of_week"]: # For each value within the column "orders of week" we want something
  if value <= 1:                                     # if the value of the row is smaller and equal to 1
    result.append("Busiest days")                    # then "Busiest day" is appended to the blank "result" list
  elif value == 3 or value == 4:                 # if the value is equal to 3 or equal to 4 
    result.append("Least busy")                      # then " Least busy" is appended to the "result" list
  else:                                              # if is neither of these conditions are met
    result.append("Regularly busy")                  # then 'Regularly busy'


In [114]:
ords_prods_merge['businest_day'] = result

In [118]:
# 3) 

ords_prods_merge.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_last_order  product_id  add_to_cart_order  reordered  \
0                    NaN         196                  1          0   
1                   15.0         196                  1          1   
2                   21.0         196                  1          1   
3                   29.0         196                  1          1   
4                   28.0         196                  1          1   

  product_name  aisle_id  department_id  prices _merge    price_range_loc  \
0         Soda        77              7     9.0   both  Mid-range product   
1         Soda        77              7     9.0   both  Mid-range product   
2         Soda        77              7     9.0   both  Mid-range product   
3         Soda        77              7     9.0   both  Mid-range product   
4         Soda        77              7     9.0   both  Mid-range product   

     businest_day businest_period_of_day  
0  Regularly busy         Average orders  
1      Least busy         Average orders  
2      Least busy            Most orders  
3      Least busy         Average orders  
4      Least busy            Most orders

# Task 3:

Check the values of this new column for accuracy. Note any observations in markdown format.

In [119]:
ords_prods_merge['businest_day'].value_counts(dropna = False)

Regularly busy    12916527
Busiest days      11864860
Least busy         7624654
Name: businest_day, dtype: int64

#### Observations: 

Now the number of the more or less busy days slip diffently comparing to the other column

- "Regularly busy" days now are 9,501,157 less than before
- "Businest days" now are 5,660,456 more 
- "Least busy" days are now 3,840,701 more

# Task 4:

When too many users make Instacart orders at the same time, the app freezes. The senior technical officer at Instacart wants you to identify the busiest hours of the day. Rather than by hour, they want periods of time labeled **“Most orders”**,  **“Average orders”**,  and **“Fewest orders”** Create a new column containing these labels called **“busiest_period_of_day”**.

1) First we need to know on which hour of the day most orders take place by printing the frequency of the **"order_hour_of_day”** column


In [80]:

ords_prods_merge['order_hour_of_day'].value_counts(dropna = False)

10    2761862
11    2736242
14    2689251
15    2662246
13    2661057
12    2618613
16    2535313
9     2454273
17    2087721
8     1718182
18    1636566
19    1258352
20     976182
7      891082
21     795658
22     634247
23     402335
6      290500
0      218786
1      115706
5       87964
2       69377
4       53245
3       51281
Name: order_hour_of_day, dtype: int64

Answer: 

- **"Most orders"** = From 9:00 to 16:00

- **"Fewest orders"** = From 23:00 to 6:00

- **"Average orders"** = From 17:00 to 22:00 and from 7:00 to 8:00


In [82]:
ords_prods_merge.shape

(32406041, 16)

2) Create the list with the categories 

In [105]:
result_2 = []

for value in ords_prods_merge['order_hour_of_day']:
    if (value >= 9 and value <= 16):                # a) From 9:00 to 16:00
        result_2.append("Most orders")              # Then "Most orders"
    elif (value >= 23) or (value <= 6):             # b) From 23:00 to 6:00
        result_2.append("Fewest Orders")            # then "fewest orders"
    elif (value >= 17 and value <= 22) or (value >= 7 and value <= 8): # From 17:00 to 22:00 and from 7:00 to 8:00
        result_2.append("Average orders")                              # Then "Average Orders"
    else:
        result_2.append()
        
#

#### Comments:

**a)** In this case, you have two conditions (value >= 9 and value <= 16) joined together with the logical operator "and"

**b)** In this case, you have two sets of conditions: (value >= 17 and value <= 22) and (value >= 7 and value <= 8). Both conditions are combined by the **or**


3) Create Column

In [106]:
ords_prods_merge['businest_period_of_day'] = result_2

# Task 5:

Print the frequency for this new column

In [107]:
ords_prods_merge['businest_period_of_day'].value_counts(dropna = False)

Most orders       21118857
Average orders     9997990
Fewest Orders      1289194
Name: businest_period_of_day, dtype: int64

# Task 7:

Export your dataframe as a pickle file (since you added new columns) and store it correctly in your “Prepared Data” folder.

In [ ]:
ords_prods_merge.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'orders_prods_merge.pkl'))
